chain/chain/src/chain.rs

```diff
@@ -3307,12 +3378,30 @@ impl Chain {
         let prev_block = self.store.get_block(block.header().prev_hash())?;
         let mut chain_update = self.chain_update();
         chain_update.apply_chunk_postprocessing(
-            me,
             &block,
             &prev_block,
             results.into_iter().collect::<Result<Vec<_>, Error>>()?,
         )?;
         chain_update.commit()?;
+
+        for shard_id in 0..self.runtime_adapter.num_shards(block.header().epoch_id())? {
+            // Update flat storage for each shard being caught up. We catch up a shard if it is tracked in the next
+            // epoch. If it is tracked in this epoch as well, it was updated during regular block processing.
+            if !self.runtime_adapter.cares_about_shard(
+                me.as_ref(),
+                block.header().prev_hash(),
+                shard_id,
+                true,
+            ) && self.runtime_adapter.will_care_about_shard(
+                me.as_ref(),
+                block.header().prev_hash(),
+                shard_id,
+                true,
+            ) {
+                self.update_flat_storage_for_block(&block, shard_id)?;
```

Comment:
Can we add a boolean argument `require_flat_storage_ready` for `update_flat_storage_for_block` so that if it is true, `update_flat_storage_for_block` will require the flat storage to be ready, instead of allowing it being created? We can set it to true in here, to make sure that when catch up happens, flat storage is not in migration.

Reply:
Mentioned this in issue for migration support: https://github.com/near/nearcore/issues/8250